src/figure/trader.c

```diff
@@ -40,16 +40,22 @@ int trader_create(void)
 
 void trader_record_bought_resource(int trader_id, resource_type resource)
 {
+    figure *f = figure_get(trader_id);
+    int is_land_trader = f->type == FIGURE_TRADE_CARAVAN || f->type == FIGURE_TRADE_CARAVAN_DONKEY || f->type == FIGURE_NATIVE_TRADER;
```

Comment:
To be honest I don't know if the Caravanserai should affect native traders, because unlike caravans they don't come from far away so they don't really benefit from the rest as much.